### DISCUSSION TITLE
datasets/pose/coco/

### DISCUSSION THREAD

**Discussion by giscus** [ultralytics/ultralytics#13816] [Docs]
# datasets/pose/coco/

Explore the COCO-Pose dataset for advanced pose estimation. Learn about datasets, pretrained models, metrics, and applications for training with YOLO.

https://docs.ultralytics.com/datasets/pose/coco/

**Comment by perry-smartsizer** [comment]
Is there any documentation around on using a YOLO Pose model when there are multiple classes, each with a separate set of keypoints? For example, how would this YAML look if we wanted to have both a "person" and a "tiger" class. Thanks!

**Reply by pderrenger** [reply]
@perry-smartsizer hi there!

Thank you for your question. To create a YAML configuration file for a dataset with multiple classes, each having a separate set of keypoints, you can follow the structure below. This example includes both "person" and "tiger" classes, each with their respective keypoints.

Here's how your YAML file might look:

```yaml
# Dataset root directory
path: ../datasets/multi-class-pose

# Train/val/test sets
train: images/train  # train images (relative to 'path')
val: images/val  # val images (relative to 'path')
test:  # test images (optional)

# Keypoints
kpt_shape: 
  person: [17, 3]  # 17 keypoints for person, with visibility
  tiger: [12, 2]  # 12 keypoints for tiger, without visibility

# Flip indices (if applicable)
flip_idx:
  person: [0, 2, 1, 4, 3, 6, 5, 8, 7, 10, 9, 12, 11, 14, 13, 16, 15]
  tiger: [1, 0, 3, 2, 5, 4, 7, 6, 9, 8, 11, 10]

# Classes dictionary
names:
  0: person
  1: tiger
```

In this configuration:
- `kpt_shape` specifies the number of keypoints and dimensions (x, y, visibility) for each class.
- `flip_idx` provides the indices for keypoint flipping, which is useful for data augmentation.
- `names` lists the class names.

Ensure your dataset follows the Ultralytics YOLO format for pose estimation, where each image has a corresponding `.txt` file with annotations in the format:

For `person` (with visibility):
```
<class-index> <x> <y> <width> <height> <px1> <py1> <p1-visibility> ... <px17> <py17> <p17-visibility>
```

For `tiger` (without visibility):
```
<class-index> <x> <y> <width> <height> <px1> <py1> ... <px12> <py12>
```

To train your model, you can use the following Python code:

```python
from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n-pose.pt")  # load a pretrained model

# Train the model
results = model.train(data="multi-class-pose.yaml", epochs=100, imgsz=640)
```

Or via CLI:

```bash
yolo detect train data=multi-class-pose.yaml model=yolov8n-pose.pt epochs=100 imgsz=640
```

For more detailed information on dataset preparation and configuration, please refer to our [Pose Estimation Datasets Documentation](https://docs.ultralytics.com/datasets/pose/).

I hope this helps! If you have any further questions, feel free to ask. 😊

**Comment by aridas96** [comment]
I have trained the yolov8n-pose.yaml model on the yolov8 dataset which I exported directly from roboflow, after training I tested it with images and videos I can't find the key point connections lines only key points are visible

**Reply by pderrenger** [reply]
@aridas96 hello,

Thank you for reaching out and for your detailed description of the issue you're encountering. It sounds like you've successfully trained your YOLOv8n-pose model, but you're not seeing the keypoint connections in your predictions.

To help you better, could you please confirm the following:

1. **Reproducible Code Example**: Ensure that you provide a minimum reproducible code example. This helps us replicate the issue on our end and investigate a solution. You can refer to our guide on creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

2. **Package Versions**: Verify that you are using the latest versions of `torch` and `ultralytics`. If not, please upgrade your packages and try again:
   ```bash
   pip install --upgrade torch ultralytics
   ```

Regarding the keypoint connections, you can visualize them by ensuring that the `show` method is correctly configured to display the connections. Here's a quick example of how you can do this:

```python
from ultralytics import YOLO

# Load the trained model
model = YOLO("path/to/your/trained/model.pt")

# Run inference on an image
results = model("path/to/your/image.jpg")

# Visualize the results with keypoint connections
for result in results:
    result.show(kpt_line=True)  # Ensure kpt_line is set to True to show connections
```

If you still face issues, please provide the specific code snippet you used for testing and any relevant details about your environment. This will help us diagnose the problem more effectively.

For more detailed information on pose estimation datasets and configurations, you can visit our [Pose Estimation Datasets](https://docs.ultralytics.com/datasets/pose/) documentation.

Looking forward to your response!

**Comment by aridas96** [comment]
I used Roboflow for annotation and downloaded the dataset in YOLOv8 TXT
format.

code snippet which i used for Training and Testing,
# For Training
from ultralytics import YOLO

model = YOLO("yolov8n-pose.yaml")  # build a new model from YAML
model = YOLO("yolov8n-pose.pt")  # load a pretrained model (recommended for
training)
model = YOLO("yolov8n-pose.yaml").load("yolov8n-pose.pt")  # build from
YAML and transfer weights

results = model.train(data="data.yaml", epochs=100, imgsz=640)

#For Testing
model = YOLO("best.pt")
results = model("path to my image.jpg")
for result in results:
    result.show(kpt_line=True)  # Ensure kpt_line is set to True to show
connections

When I try to test my image I am getting the key points at the exact
position as I annotated but the key point connections(lines) are missing.
I have also attached the data.yaml file which I used for training.

**Reply by glenn-jocher** [reply]
Hi @aridas96,

Thank you for providing detailed information about your issue. It sounds like you've successfully trained your YOLOv8n-pose model, but you're encountering a problem with visualizing keypoint connections during inference.

To assist you better, could you please confirm the following:

1. **Reproducible Code Example**: Ensure that you provide a minimum reproducible code example. This helps us replicate the issue on our end and investigate a solution. You can refer to our guide on creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

2. **Package Versions**: Verify that you are using the latest versions of `torch` and `ultralytics`. If not, please upgrade your packages and try again:
    ```bash
    pip install --upgrade torch ultralytics
    ```

Regarding the keypoint connections, it seems like you are already using the `kpt_line=True` parameter, which is correct. Here’s a quick example to ensure everything is set up properly:

```python
from ultralytics import YOLO

# Load the trained model
model = YOLO("best.pt")

# Run inference on an image
results = model("path/to/your/image.jpg")

# Visualize the results with keypoint connections
for result in results:
    result.show(kpt_line=True)  # Ensure kpt_line is set to True to show connections
```

If you still face issues, please provide the specific code snippet you used for testing and any relevant details about your environment. This will help us diagnose the problem more effectively.

For more detailed information on pose estimation datasets and configurations, you can visit our [Pose Estimation Datasets](https://docs.ultralytics.com/datasets/pose/) documentation.

Looking forward to your response!

**Comment by xhf3571** [comment]
Hi,

I've noticed some inconsistencies in the reported numbers of images in the COCO-Pose dataset across different sources. For instance, some papers mention that the train2017 set includes 57K images, whereas the val2017 and test-dev2017 sets consist of 5K and 20K images respectively. However, the dataset I downloaded has over 64,000 images in the training set and only about 2,600 images in the validation set.

Could you please clarify the official numbers of images used in the COCO-Pose dataset? Knowing the exact figures would be very helpful for my research.

Thank you for your assistance.

Best regards,

**Reply by xhf3571** [reply]
I recently redownloaded the COCO-pose using the coco-pose.yaml configuration. I noticed that in the coco-pose/labels/val2017 directory, there are only 2346 .txt files, whereas the coco-pose/images/val2017 directory contains 5000 images. Is this discrepancy expected?

Additionally, I observed that coco-pose/labels/train2017 contains 56599 .txt files, but I encountered a failure while downloading the train2017 images. I plan to manually download these images and match them with the 56599 .txt files. Would this approach be appropriate?

Thank you for your assistance.

**Reply by glenn-jocher** [reply]
@xhf3571 hi there,

Thank you for your detailed observation and for bringing this to our attention.

The discrepancy you're noticing between the number of images and labels in the COCO-Pose dataset is indeed expected. The COCO dataset, including its pose estimation subset, often has more images than annotations because not all images contain the objects or keypoints of interest.

### Val2017 Set
- **Images**: 5000
- **Annotations**: 2346

### Train2017 Set
- **Images**: 118,287 (as per the full COCO dataset)
- **Annotations**: 56,599

For the **val2017** set, the 5000 images include those without keypoint annotations, which explains why there are fewer `.txt` files. Similarly, for the **train2017** set, the number of images and annotations can differ due to the same reason.

### Downloading and Matching Images
Your approach to manually download the images and match them with the 56,599 `.txt` files is appropriate. Ensure that the images are correctly placed in the directory structure as specified in the `coco-pose.yaml` file.

```yaml
path: ../datasets/coco-pose
train: train2017.txt
val: val2017.txt
test: test-dev2017.txt
```

### Additional Resources
For more detailed information and to ensure you're using the latest dataset configuration, please refer to the [COCO-Pose Dataset Documentation](https://docs.ultralytics.com/datasets/pose/coco/).

If you encounter any further issues or have additional questions, please provide a minimum reproducible example of your setup and code. This will help us assist you more effectively. You can find guidelines for creating a reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example).

Thank you for your understanding and cooperation. Happy training! 🚀

**Comment by xhf3571** [comment]
### Issue: Error While Running Inference on COCO Test Set

Hello,

I saw your response and code regarding the COCO test set, and I tried running the following code:

```python
from ultralytics import YOLO
import json

# Load the model
model = YOLO("/my/path/best.pt")

# Run inference on the test set
results = model.val(data="coco-pose.yaml", split="test")

# Convert results to COCO format
coco_results = []
for result in results:
    for box in result.boxes:
        coco_results.append({
            "image_id": result.image_id,
            "category_id": box.cls,
            "bbox": box.xywh.tolist(),
            "score": box.conf,
            "keypoints": box.keypoints.tolist()
        })

# Save results to JSON file
with open("coco_results.json", "w") as f:
    json.dump(coco_results, f)
```

However, I encountered the following error:

```
  raise FileNotFoundError(f"{self.prefix}Error loading data from {img_path}\n{HELP_URL}") from e
FileNotFoundError: val: Error loading data from /mnt/workspace/ultralytics-8.1.45/ultralytics/cfg/datasets/test-dev2017.txt
See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
```

Could you please tell me the correct steps to resolve this issue? Additionally, what should the output file format be after running the code? What contents should it include? Lastly, should this generated file be submitted to this URL?
[https://codalab.lisn.upsaclay.fr/competitions/7403#participate-submit_results](https://codalab.lisn.upsaclay.fr/competitions/7403#participate-submit_results)

Thank you!

---

**Reply by glenn-jocher** [reply]
@xhf3571 hello,

It looks like the error is due to a missing or incorrect path to the `test-dev2017.txt` file. Ensure that your dataset directory structure matches the paths specified in your `coco-pose.yaml` file. Verify that the `test-dev2017.txt` file exists in the correct location.

Regarding the output file format, your JSON should include the image ID, category ID, bounding box coordinates, confidence score, and keypoints. This format is suitable for submission to the COCO evaluation server, as specified in the competition guidelines.

For more details on dataset formatting, you can refer to the [Ultralytics documentation](https://docs.ultralytics.com/datasets/pose/coco8-pose/).

Best of luck!

**Reply by xhf3571** [reply]
以下是英文简洁版的提问：
I adjusted and ran the code, but it still reports an error saying no labels found. I think it is normal for the test set to have no labels. The error is as follows:
```markdown
WARNING ⚠️ No labels found in /mnt/workspace/ultralytics-8.1.45/ultralytics/assets/test2017.cache, training may not work correctly. See https://docs.ultralytics.com/datasets/detect for dataset formatting guidance.
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95)     Pose(P          R      mAP50  mAP50-95): 100%|██████████| 1268/1268 [00:57<00:00, 21.92it/s]
                   all      20288          0          0          0          0          0          0          0          0          0
WARNING ⚠️ no labels found in pose set, can not compute metrics without labels
Speed: 0.1ms preprocess, 1.3ms inference, 0.0ms loss, 0.6ms postprocess per image
Results saved to runs/pose/val6
Traceback (most recent call last):
  File "/mnt/workspace/ultralytics-8.1.45/ultralytics/coco-test.py", line 12, in <module>
    for result in results:
TypeError: 'PoseMetrics' object is not iterable
```

Thank you!

**Reply by xhf3571** [reply]
Hi there,
I've looked through other responses and tried the following command:
```markdown


```bash
(yolov8Pose) root@dsw-405310-8475c8cfcb-zbvfs:/mnt/workspace/ultralytics-8.1.45/ultralytics# python detect --source /mnt/workspace/ultralytics-8.1.45/ultralytics/assets/test2017 --weights /mnt/workspace/ultralytics-8.1.45/ultralytics/runs/train/C2F_CCA/weights/best.pt --save-json
```

However, I encountered an error:

```
python: can't open file '/mnt/workspace/ultralytics-8.1.45/ultralytics/detect': [Errno 2] No such file or directory
```

Could you please advise which directory I should be running this command from?

Additionally, I tried running the following command:

```bash
python predict model=/mnt/workspace/ultralytics-8.1.45/ultralytics/runs/train/C2F_CCA/weights/best.pt data=coco-pose.yaml split=test
```

But I received the same type of error:

```
python: can't open file '/mnt/workspace/ultralytics-8.1.45/ultralytics/predict': [Errno 2] No such file or directory
```

Could you help clarify where these scripts are located or if there is a step I might be missing?

Thanks for your help!
```

**Reply by xhf3571** [reply]
If the operation is successful, should a JSON file be generated? I noticed that the final submission to the website https://competitions.codalab.org/competitions/12061#participate is a ZIP file. Could you please let me know what else should be included in this ZIP file?

**Reply by glenn-jocher** [reply]
@xhf3571 yes, a JSON file should be generated if the operation is successful. For the final submission to the competition, you need to include this JSON file in a ZIP archive. The ZIP file should contain the JSON results formatted according to the competition guidelines. Make sure to follow the specific instructions provided on the competition's submission page to ensure your results are correctly evaluated.

**Reply by xhf3571** [reply]
I've posted my specific issue on the GitHub project, and the issue link is: https://github.com/ultralytics/ultralytics/issues/14510. I really need your help as I've been stuck for 3 days now. Thank you so much!

**Reply by glenn-jocher** [reply]
@xhf3571 thank you for sharing the issue link. Please ensure you are using the latest version of the Ultralytics package, as updates often resolve such issues. The error indicates that no labels were found, which is expected for the test set. However, the `PoseMetrics` object is not iterable, which suggests a possible bug or misuse.

For running inference on the test set, you should use the `predict` mode instead of `val`. The `predict` mode does not require labels and should generate the JSON file you need. Here's a minimal example:

```python
from ultralytics import YOLO

# Load the model
model = YOLO("/my/path/best.pt")

# Run inference on the test set
results = model.predict(data="coco-pose.yaml", split="test")

# Process and save results as needed
```

Ensure the paths in your `coco-pose.yaml` are correct and point to the appropriate test set files. Once you have the JSON results, you can zip the file and submit it as required by the competition guidelines. If the issue persists, please update the GitHub issue with any new error messages or behavior observed.

**Reply by xhf3571** [reply]
Thank you for your response. I have now switched to the latest version of YOLOv8, version 8.2.59, and am running the following code:
```markdown
```python
from ultralytics import YOLO
import json

# Load the model
model = YOLO("/root/lanyun-tmp/ultralytics-8.2.59/ultralytics/runs/train/C2F_CCA/weights/best.pt")

# Run inference on the test set
results = model.predict(source="/root/lanyun-tmp/ultralytics-8.1.45/ultralytics/assets/test2017")

# Convert results to COCO format
coco_results = []
for result in results:
    for box in result.boxes:
        coco_results.append({
            "image_id": result.image_id,
            "category_id": box.cls,
            "bbox": box.xywh.tolist(),
            "score": box.conf,
            "keypoints": box.keypoints.tolist()
        })

# Save results to JSON file
with open("coco_results.json", "w") as f:
    json.dump(coco_results, f)
```

I encountered the following error:

```
image 40665/40670 /root/lanyun-tmp/ultralytics-8.1.45/ultralytics/assets/test2017/000000581864.jpg: 480x640 2 persons, 16.0ms
image 40666/40670 /root/lanyun-tmp/ultralytics-8.1.45/ultralytics/assets/test2017/000000581872.jpg: 480x640 (no detections), 15.5ms
image 40667/40670 /root/lanyun-tmp/ultralytics-8.1.45/ultralytics/assets/test2017/000000581897.jpg: 448x640 (no detections), 15.8ms
image 40668/40670 /root/lanyun-tmp/ultralytics-8.1.45/ultralytics/assets/test2017/000000581908.jpg: 640x640 (no detections), 18.1ms
image 40669/40670 Traceback (most recent call last):
  File "/root/autodl-tmp/ultralytics-8.2.59/ultralytics/coco-test.py", line 17, in <module>
    "image_id": result.image_id,
  File "/root/autodl-tmp/ultralytics-8.2.59/ultralytics/utils/__init__.py", line 160, in __getattr__
    raise AttributeError(f"'{name}' object has no attribute '{attr}'. See valid attributes below.\n{self.__doc__}")
AttributeError: 'Results' object has no attribute 'image_id'. See valid attributes below.
ValueError: not enough values to unpack (expected 4, got 1)
```

I am using the new version of YOLOv8, and it seems to predict correctly, but I am still unable to generate the JSON file for evaluation on the COCO test set. Could you please provide me with the correct and complete code to generate this? I have been working on this for four days now.

Thank you and look forward to your reply

**Reply by glenn-jocher** [reply]
@xhf3571 thank you for updating to the latest version. The error you're encountering indicates that the `Results` object does not have an `image_id` attribute. Instead, you should access the image path and extract the image ID from it. Here's a revised approach:

```python
from ultralytics import YOLO
import json
import os

# Load the model
model = YOLO("/root/lanyun-tmp/ultralytics-8.2.59/ultralytics/runs/train/C2F_CCA/weights/best.pt")

# Run inference on the test set
results = model.predict(source="/root/lanyun-tmp/ultralytics-8.1.45/ultralytics/assets/test2017")

# Convert results to COCO format
coco_results = []
for result in results:
    image_id = int(os.path.basename(result.path).split('.')[0])
    for box in result.boxes:
        coco_results.append({
            "image_id": image_id,
            "category_id": box.cls,
            "bbox": box.xywh.tolist(),
            "score": box.conf,
            "keypoints": box.keypoints.tolist()
        })

# Save results to JSON file
with open("coco_results.json", "w") as f:
    json.dump(coco_results, f)
```

This should resolve the attribute error and correctly generate the JSON file. Once you have the JSON file, you can zip it and submit it to the competition as specified. If you encounter further issues, please update the GitHub issue with the new error details.

**Reply by xhf3571** [reply]
Thank you for your response. I've run into an issue when executing the new code:
```markdown


```python
from ultralytics import YOLO
import json
import os

# Load the model
model = YOLO("/root/lanyun-tmp/ultralytics-8.2.59/ultralytics/runs/train/C2F_CCA/weights/best.pt")

# Run inference on the test set
results = model.predict(source="/root/lanyun-tmp/ultralytics-8.1.45/ultralytics/assets/test0003")

# Convert results to COCO format
coco_results = []
for result in results:
    image_id = int(os.path.basename(result.path).split('.')[0])
    for box in result.boxes:
        coco_results.append({
            "image_id": image_id,
            "category_id": box.cls,
            "bbox": box.xywh.tolist(),
            "score": box.conf,
            "keypoints": box.keypoints.tolist()
        })

# Save results to JSON file
with open("coco_results.json", "w") as f:
    json.dump(coco_results, f)
```

The error I'm encountering is:
```
Traceback (most recent call last):
  File "coco-test.py", line 21, in <module>
    "keypoints": box.keypoints.tolist()
  File "/root/lanyun-tmp/ultralytics-8.2.59/ultralytics/utils/__init__.py", line 162, in __getattr__
    raise AttributeError(f"'{name}' object has no attribute '{attr}'. See valid attributes below.\n{self.__doc__}")
AttributeError: 'Boxes' object has no attribute 'keypoints'. See valid attributes below.

```
Thank you and look forward to your reply

**Reply by glenn-jocher** [reply]
@xhf3571 it looks like the `Boxes` object does not have a `keypoints` attribute. This is likely because the model you are using is not a pose estimation model or the predictions do not include keypoints. Ensure you are using a pose-specific model like `yolov8n-pose.pt`. If the problem persists, please verify that your model and dataset are correctly configured for pose estimation. If you need further assistance, please update the issue with any new details.

**Reply by xhf3571** [reply]
Thank you for your reply. I used the official `yolov8n-pose.pt` but still encountered the same error. I referred to the issue [link](https://github.com/ultralytics/ultralytics/issues/4802) and raised a question in my own issue [link](https://github.com/ultralytics/ultralytics/issues/14510). I am very confused about which method to use to correctly generate the JSON evaluation file. This seems like a small task, but I have been struggling with it for 5 days. Please help me, I would be extremely grateful.

Thank you very much, and I look forward to your response.

**Reply by glenn-jocher** [reply]
Thank you for your patience. Since you're using the official `yolov8n-pose.pt` model and still encountering issues, please ensure you have the latest version of the Ultralytics package. The error suggests that the `Boxes` object does not contain `keypoints`. Verify that your model is correctly configured for pose estimation and that your dataset paths are accurate. If the problem persists, please update your GitHub issue with any new details or errors. This will help us provide more targeted assistance.

**Comment by LeBron-Jian** [comment]
hello， I noticed that to train the pose model using the coco dataset, I can use the following code snippet:
```from ultralytics import YOLO

# Load a model
model = YOLO("yolov8n-pose.pt")  # load a pretrained model (recommended for training)

# Train the model
results = model.train(data="coco-pose.yaml", epochs=100, imgsz=640, device=[0,1])
```

But I tested the effect of the given pre-trained model(yolov8n-pose.pt), and the results are as follows:
![coco_res](https://github.com/user-attachments/assets/6cbe724c-cac2-4024-8e85-2ad39c279392)

Then the effect of the pre-trained model has reached the given indicators. So what is the point of using the pre-trained model for training directly, and can it still improve the accuracy?
![official_res](https://github.com/user-attachments/assets/c4563d22-7432-48c0-9716-ffaec2b05a84)

therefore, if i want to reproduce the effect of yolo-pose on the coco dataset, how should I do it?

**Reply by glenn-jocher** [reply]
@LeBron-Jian using a pretrained model like YOLOv8n-pose provides a strong starting point, allowing you to fine-tune on your specific data to potentially improve accuracy. To reproduce the effects on the COCO dataset, ensure you're using the correct dataset configuration and training parameters as outlined in the `coco-pose.yaml`. For more details, refer to the COCO-Pose documentation.

**Reply by LeBron-Jian** [reply]
@glenn-jocher  hi, thanks for your reply.  I went through the documentation https://docs.ultralytics.com/datasets/pose/coco/,。
I used the original configuration file and did not use the pre-trained model. Can this reproduce the effect of yolo-pose on the coco dataset?
the code is as follows:
```
    #  build a new model from YAML
    model = YOLO("cfg/models/v8/yolov8-pose.yaml")  
    # Train the model
    model.train(data='cfg/datasets/coco-pose.yaml', batch=16, workers=8, device=[0, 1])

```

**Reply by glenn-jocher** [reply]
To reproduce the YOLO-Pose results on the COCO dataset, ensure you're using the correct `coco-pose.yaml` configuration and consider starting with a pretrained model for better initial performance. Fine-tuning on your specific data can help improve accuracy.

**Reply by LeBron-Jian** [reply]
> To reproduce the YOLO-Pose results on the COCO dataset, ensure you're using the correct `coco-pose.yaml` configuration and consider starting with a pretrained model for better initial performance. Fine-tuning on your specific data can help improve accuracy.

hi， I trained the pose model  on my own small dataset based on the pre-trained model. my training set is about 200 pictures and the validation set is 40 pictures. the scene is a very simple scene with no occlusion and pure background. I found that the mAP50~95 of both bbox and keypoint are very high, but the keypoints of the actual test pictures are not so accurate. why is this?
![res_pose](https://github.com/user-attachments/assets/fbbabc89-e541-40a7-9b76-a701afb914c3)
the only thing i can think of is the setting of sigmas used in the calculation of metrics. the sigmas i gave is the default `torch.ones(nkpt)/nkpt` in the code

**Reply by glenn-jocher** [reply]
High mAP but inaccurate keypoints can result from overfitting, especially with a small dataset. Consider increasing your dataset size or using data augmentation to improve generalization. Adjusting sigma values might help, but focus on data diversity first.

**Reply by LeBron-Jian** [reply]
Hello, according to your suggestion, I diversified my own small hand dataset. Then I trained it based on the pre-trained model. I found a strange phenomenon, it is not that the larger the model, the better the effect. (Based on the experience of object detection, the X model is always better than the L model, M model, and S model), but the mAP of the bbox here is relatively stable. As for the mAP of Pose, the mAP of the X model reached 0.888, but the L model only had 0.664, the M model had 0.748, the S model had 0.928, and the nano model had an accuracy of 0.914. This conclusion was obtained when I controlled the training set and the validation set, and epochs=100. This makes me a little confused, why is it? Does it mean that the fine-tuning of the pre-trained model for a small dataset is not that the larger the model, the better the effect?

**Reply by LeBron-Jian** [reply]
Is it because the larger the model, the harder it is to converge? Do I need more epochs?

**Reply by glenn-jocher** [reply]
@LeBron-Jian model size doesn't always correlate with better performance, especially on small datasets. Smaller models might generalize better due to fewer parameters. Consider experimenting with different architectures and hyperparameters to find the best fit for your specific dataset.

**Comment by xcf1** [comment]
Hello, I used the trained model to infer the test-dev dataset of coco2017, and converted the inference results into the json format required by coco official. My json file format is like [{"image_id": 16, "category_id": 1, "keypoints": [310, 382, 0, 377, 404, 0, 438, 412, 0, 444, 0, 0, 148, 452, 0, 445, 0, 0, 19, 499, 0, 476, 424, 0, 455, 503, 0, 474, 373, 0, 393, 391, 0, 473, 354, 0, 420, 495, 0, 479, 463, 0, 478, 365, 0, 479, 471, 0, 477, 317, 0], "score": 0.962377}, {"image_id": 16, "category_id": 1, "keypoints": [21, 341, 0, 267, 0, 0, 23, 0, 0, 22, 0, 0, 7, 0, 0, 78, 0, 0, 13, 0, 0, 167, 0, 0, 53, 0, 0, 237, 0, 0, 53, 0, 0, 212, 0, 0, 75, 14, 0, 323, 7, 0, 240, 9, 0, 292, 0, 0, 223, 0, 0], "score": 0.233321}, {"image_id": 57, "category_id": 1, "keypoints": [443, 223, 0, 556, 263, 0, 629, 268, 0, 618, 259, 0, 592, 280, 0, 568, 256, 0, 405, 290, 0, 638, 256, 0, 635, 306, 0, 632, 284, 0, 612, 326, 0, 624, 325, 0, 597, 288, 0, 639, 265, 0, 638, 308, 0, 638, 281, 0, 638, 300, 0], "score": 0.988164},...........], but after uploading to the official website, I still got an error: AssertionError: Results do not correspond to current coco set. Could you please tell me the reason? Can you provide an accurate json file format, or give me a code that converts the predicted results into the correct json file? Thank you very much.

**Reply by glenn-jocher** [reply]
@xcf1 thank you for your question. The AssertionError typically occurs when your results file doesn't match COCO's expected format for keypoints. For COCO test-dev2017 submissions, please ensure your JSON follows the exact format shown in our [COCO-Pose evaluation documentation](https://docs.ultralytics.com/datasets/pose/coco/#evaluation). The keypoints array should contain 17 keypoints in (x,y,visibility) format where visibility=2 for visible points. 

Here's a minimal validation script to generate correct JSON:

```python
from ultralytics import YOLO

model = YOLO('yolo11n-pose.pt')
model.val(data='coco-pose.yaml', split='test-dev', save_json=True)
```

This will automatically generate a properly formatted `results.json` in your runs/val directory. For further assistance, please review the [COCO evaluation guidelines](https://cocodataset.org/#upload).

**Reply by xcf1** [reply]
Thank you very much for your reply. I have obtained the results.json file, but when I upload its compressed package to the COCO official website, the webpage displays a file loading failure. What could be the reason for this? My compressed file size is 238MB, while the COCO official website specifies a size of about 200MB. Also, the img_id in this JSON file is arranged in descending order. Is there a requirement for the order of IDs? I have also noticed an issue: for the image with img_id 16, the corresponding generated txt file contains many lines of data, even though there are only two people in the image, suggesting a mismatch in data. Other images seem to be normal.

**Reply by xcf1** [reply]
this is isuue_img
![DC3A0EAFDC982B6146C9E090934A69AF](https://github.com/user-attachments/assets/cb6e5df6-d218-4ef0-8365-1a754ab7cedd)

**Reply by glenn-jocher** [reply]
Thank you for the follow-up. The file size discrepancy and ID ordering issues typically occur when not using the official test-dev2017 split. For COCO evaluation server submissions, please ensure you're validating with `split='test-dev'` in your validation command to generate results that strictly correspond to the 20,288 test images. The JSON should auto-generate with correct IDs and sizing when using: 

```python
model.val(data='coco-pose.yaml', split='test-dev', save_json=True)
```

For confidence threshold adjustments, you can add `conf=0.01` (COCO standard) to reduce detections. If issues persist, please verify your dataset YAML points to test-dev2017.txt as shown in the [COCO-Pose dataset configuration](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/datasets/coco-pose.yaml).

**Reply by xcf1** [reply]
Hello, thank you for your reply. I have increased the conf value to reduce the detection size, and now it can be uploaded to the COCO official website for testing. However, I still get the error “AssertionError: Results do not correspond to current coco set.” The content of the JSON file generated by the code you provided is as shown in the image below. What could be the reason for this? I have been struggling with this issue for a long time, and it still persists.
![97A55D72D352687CFF40DCF131CB43A1](https://github.com/user-attachments/assets/569889e2-d749-4516-a88f-b30ac8182a33)

**Reply by glenn-jocher** [reply]
The "AssertionError" typically indicates mismatched image IDs between your results and COCO's test-dev2017 set. Please ensure you're validating with the exact command:  
`model.val(data='coco-pose.yaml', split='test-dev', save_json=True)`  
This guarantees proper ID alignment. For direct verification, compare your JSON's image IDs with those in the [test-dev2017.txt](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/datasets/coco-pose.yaml#L13) from our dataset YAML.

**Reply by xcf1** [reply]
![AF922C9058DD5C99E623F42C1BE97A4C](https://github.com/user-attachments/assets/1f1b59f0-aa04-4ef4-b6bb-c314cd38cb95)
Hello, I have already ensured that my results correspond to the images in the COCO2017 test-dev dataset. When I submitted a small subset of my model's inference results in JSON format to the COCO official service platform, it worked and I received evaluation results. However, after adding more images' JSON results, the platform threw an error: "AssertionError: Results do not correspond to current COCO set". I have confirmed that the additional images, such as the one with ID 426923, are indeed in the COCO2017 test-dev dataset. Could you please tell me why this is happening? Thank you very much.
hank you very much.
![381E1B64E3D5EE5DEDE18943B95F258A](https://github.com/user-attachments/assets/2f0273a1-d1d0-4268-bbe8-a0a9f5c2dab2)

![A7085D6993C6D36C72983595CEFB185C](https://github.com/user-attachments/assets/f2c13a5b-f5f8-4e0c-afca-c0e14348fce9)

**Reply by glenn-jocher** [reply]
The AssertionError typically occurs when your JSON contains image IDs not present in the official COCO test-dev2017 split (20,288 images). Please verify that *all* image IDs in your results match those in the [test-dev2017.txt](https://github.com/ultralytics/ultralytics/blob/main/ultralytics/cfg/datasets/coco-pose.yaml#L13) list from our dataset YAML. For validation, try loading your JSON with the COCO API's `loadRes()` function to detect mismatches.

**Reply by xcf1** [reply]
thank you very much，the isuue is solved

**Reply by glenn-jocher** [reply]
Great to hear your issue is resolved! For future users encountering similar COCO evaluation challenges, ensure your dataset YAML points to the correct test split and use the official `model.val()` method to generate results. Let us know if any other questions arise.
